cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -427,6 +429,162 @@ class self_comparator {
   std::shared_ptr<preprocessed_table> d_t;
 };
 
+template <typename Nullate>
+class two_table_device_row_comparator_adapter {
+  friend class two_table_comparator;
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ bool operator()(lhs_index_type const lhs_index,
+                             rhs_index_type const rhs_index) const noexcept
+  {
+    return comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));
+  }
+
+  /**
+   * @brief Checks whether the row at `rhs_index` in the `rhs` table compares
+   * lexicographically less than the row at `lhs_index` in the `lhs` table.
+   *
+   * @param rhs_index The index of row in the `rhs` table to examine
+   * @param lhs_index The index of the row in the `lhs` table to examine
+   * @return `true` if row from the `rhs` table compares less than row in the `lhs` table
+   */
+  __device__ bool operator()(rhs_index_type const rhs_index,
+                             lhs_index_type const lhs_index) const noexcept
+  {
+    // TODO: "not lhs < rhs" isn't quite the same as "rhs < lhs". The case of
+    // equality returns true for operator(rhs, lhs), while operator(lhs, rhs)
+    // returns false. This would have to be handled at a lower level, if it
+    // matters. Do we just document that this means "rhs <= lhs"?
+    return not comp(static_cast<cudf::size_type>(lhs_index),
+                    static_cast<cudf::size_type>(rhs_index));
```

Comment:
I think we should document this carefully. In addition, if we care about ordering then we should overcome this somehow. If we are in a stable sort, comparing `rhs < lhs` will be true if `lhs == rhs` (as comparing `lhs < rhs == false`), and sorting using the result of `rhs < lhs` will incorrectly swap the elements.

Reply:
Yep, that's exactly the case that led me to write this comment. I'm not ready to address this problem yet, until I have a better grasp on which Thrust algorithms support comparators accepting two different types. `thrust::merge`, for instance, will not.

Separating the comparators into two APIs may not have the effect that we want here. We may need to handle each of the `operator()` overloads differently for the case of equality. I need to do some more analysis before I am confident in what we should expect from the behavior here.

Reply:
This might not even be the right logic. We might want to remove the `not` and always check `lhs[i] < rhs[j]` regardless of the order in which the indices were passed. In that case, the alternative of `rhs[i] < lhs[j]` should be evaluated by flipping the input tables. I got a draft of `cudf::search` working but I'm not yet certain of how the `operator()(rhs_index_type, lhs_index_type)` should behave.

Reply:
Essentially that is just the `LESS` operator. So we should want to have the *exact* result of the `LESS` comparison. Internal implementation details can vary in any way.

Reply:
Would it help if the comparator returned weak state instead of bool? Afaik we have a requirement to return the weak state for other purposes so we'll have to make the change sooner or later.

Reply:
Yes! I am working with @rwlee to untangle how this PR should interact with #9452. The weak ordering changes are one aspect of that, which would help eliminate this ambiguity.

Reply:
I'm not sure how much weak order can help on this. Many external algorithms (like `thrust::sort`) rely on strict weak ordering.

Reply:
@ttnghia, When the `device_row_comparator` returns `weak_order`, we can do more than just `comp()` and `not comp()`. Which means this function can return `(comp(lhs, rhs) == GREATER)`.

How that will work when the user **wants** the `weak_order` as a return remains to be figured out. But I'm sure we'll figure out something by adding layers.